app/src/renderer/vuex/modules/ledger.js

```diff
@@ -0,0 +1,119 @@
+import * as Sentry from "@sentry/browser"
+import { App, comm_u2f } from "ledger-cosmos-js"
+import { createCosmosAddress } from "../../scripts/wallet.js"
+
+// TODO: discuss TIMEOUT value
+const TIMEOUT = 50 // seconds to wait for user action on Ledger
+const HDPATH = [44, 118, 0, 0, 0]
+
+export default () => {
+  const emptyState = {
+    error: null,
+    app: null, // Cosmos ledger app instance
+    isConnected: false,
+    pubKey: null, // 33 bytes; used for broadcasting signed txs
+    uncompressedPubKey: null, // 65 bytes; not used currently
+    version: null // Cosmos app version
+  }
+  const state = {
+    ...emptyState,
+    externals: { createCosmosAddress, App, comm_u2f } // for testing
+  }
+  const mutations = {
+    setLedger(state, app) {
+      state.app = app
+    },
+    setLedgerCosmosVersion(state, version) {
+      state.version = version
+    },
+    setLedgerPubKey(state, pubKey) {
+      state.pubKey = pubKey
+    },
+    setLedgerUncompressedPubKey(state, uncompressedPubKey) {
+      state.uncompressedPubKey = uncompressedPubKey
+    },
+    setLedgerConnection(state, isConnected) {
+      state.isConnected = isConnected
+    },
+    setLedgerError(state, error) {
+      state.error = error
+    }
+  }
+
+  const actions = {
+    resetSessionData({ rootState }) {
+      rootState.ledger = {
+        ...JSON.parse(JSON.stringify(emptyState)),
+        externals: state.externals
+      }
+    },
+    /* TODO: Create a function to detect ledger is connected (i.e unlocked with
+      password and on Home app) */
+    async connectLedgerApp({ commit, dispatch, state }) {
+      try {
+        const comm = await state.externals.comm_u2f.create_async(TIMEOUT, true)
+        const app = new state.externals.App(comm)
+        commit(`setLedger`, app)
+        await dispatch(`getLedgerCosmosVersion`)
+        commit(`setLedgerConnection`, true)
+        await dispatch(`getLedgerPubKey`)
+        const address = state.externals.createCosmosAddress(state.pubKey)
+        await dispatch(`signIn`, { sessionType: `ledger`, address })
+      } catch (error) {
+        commit(`notifyError`, {
+          title: `Error connecting to Ledger`,
+          body: error.message
+        })
+        Sentry.captureException(error)
+        commit(`setLedgerError`, error.message)
+      } finally {
+        return !state.error
+      }
+    },
+    async getLedgerCosmosVersion({ commit, dispatch, state }) {
```

Comment:
most of these functions do not need to be action, maybe we could put them into a `ledeger` script

Reply:
that would be a step towards a more universal library - right?

Reply:
yes, but also these functions are not really manipulating the state and could be written as pure functions. but maybe we do this refactor another time.